File: 자료구조 부수기/BOJ_19638.java
package BOJ_19638;

import java.io.*;
import java.util.Collections;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class BOJ_19638 {

    // https://www.acmicpc.net/problem/19638
    // input
    private static BufferedReader br;

    // variables
    private static int N, H, T;
    private static PriorityQueue<Integer> pque;


    public static void main(String[] args) throws IOException {
        System.setIn(new FileInputStream("C:\\Users\\bigyo\\Desktop\\알고리즘\\JavaAlgorithm\\src\\BOJ_19638\\res.txt"));
        br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        input();

        bw.write(solve());
        bw.close();
    } // End of main()

    private static String solve() {
        StringBuilder sb = new StringBuilder();

        int ans = 0;
        while (T-- > 0) {
            int cur = pque.poll();
            if (cur < H) {
                pque.offer(cur);
                break;
            }

            int temp = cur / 2;
            if (temp <= 0) {
                pque.offer(1);
            } else {
                pque.offer(temp);
            }
            ans++;
        }

        if (pque.peek() < H) {
            sb.append("YES\n").append(ans);
        } else {
            sb.append("NO\n").append(pque.peek());
        }

        return sb.toString();
    } // End of solve()

    private static void input() throws IOException {
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        H = Integer.parseInt(st.nextToken());
        T = Integer.parseInt(st.nextToken());

        pque = new PriorityQueue<>(Collections.reverseOrder());
        for (int i = 0; i < N; i++) {
            pque.offer(Integer.parseInt(br.readLine()));
        }
    } // End of input()
} // End of Main class

File: 그리디 부수기/BOJ 19638.kt
package BOJ_19638

import java.io.File
import java.util.PriorityQueue
import java.util.StringTokenizer


// input
private var br = System.`in`.bufferedReader()

// variables
private var N = 0
private var H = 0L
private var T = 0
private lateinit var que: PriorityQueue<Long>

fun main() {
    val path = "C:\\Users\\bigyo\\Desktop\\알고리즘\\KotlinAlgo\\src\\main\\kotlin\\BOJ_19638\\res.txt"
    br = File(path).bufferedReader()
    val bw = System.out.bufferedWriter()

    input()

    bw.write(solve())
    bw.close()
} // End of main()

private fun solve(): String {
    val sb = StringBuilder()

    var count = 0
    var flag = false
    while (T-- > 0) {
        val cur = que.peek()
        if (cur == 1L || cur < H) {
            flag = true
            break
        }
        que.poll()
        que.offer(cur / 2)
        count++
    }

    if (que.peek() < H) {
        sb.append("YES").append('\n').append(count)
    } else {
        sb.append("NO").append('\n').append(que.poll())
    }

    return sb.toString()
} // End of solve()

private fun input() {
    StringTokenizer(br.readLine()).run {
        N = nextToken().toInt()
        H = nextToken().toLong()
        T = nextToken().toInt()
    }

    que = PriorityQueue(reverseOrder())
    repeat(N) {
        que.offer(br.readLine().toLong())
    }
} // End of input()